src/coreclr/jit/emitxarch.cpp

```diff
@@ -979,7 +979,7 @@ bool emitter::AreFlagsSetToZeroCmp(regNumber reg, emitAttr opSize, GenCondition
     // Certain instruction like and, or and xor modifies exactly same flags
     // as "test" instruction.
     // They reset OF and CF to 0 and modifies SF, ZF and PF.
-    if (DoesResetOverflowAndCarryFlags(lastIns))
+    if (DoesResetOverflowAndCarryFlags(lastIns) && DoesWriteSignFlag(lastIns))
```

Comment:
Is this "sufficient"? The comment seems to indicate we want to exactly match what `test` is looking for and this isn't checking that `PF` or `ZF` are written.

The general `AreFlagsSetToZeroCmp` is meant to determine if we can drop `test x, x`; and the flags may be incorrect for ZF/PF if we didn't check them as well.

Reply:
-- Given our current instructions, we only have 7 instructions which reset both overflow and carry:
* `or     - Resets_OF | Resets_SF    | Writes_ZF | Resets_AF    | Resets_PF    | Resets_CF`
* `and    - Resets_OF | Undefined_SF | Writes_ZF | Undefined_AF | Undefined_PF | Resets_CF`
* `xor    - Resets_OF | Writes_SF    | Writes_ZF | Undefined_AF | Undefined_PF | Resets_CF`
* `test   - Resets_OF | Writes_SF    | Writes_ZF | Undefined_AF | Writes_PF    | Resets_CF`
* `andn   - Resets_OF | Writes_SF    | Writes_ZF | Undefined_AF | Writes_PF    | Resets_CF`
* `bestr  - Resets_OF | Writes_SF    | Writes_ZF | Undefined_AF | Writes_PF    | Resets_CF`
* `popcnt - Resets_OF | Writes_SF    | Writes_ZF | Undefined_AF | Writes_PF    | Resets_CF`


So:
* `or` and `and` both violate the `SF` expectation of `test`
* `or`, `and`, and `xor` all violate the `PF` expectation of `test`
* there is no `AF` expectation of test

I think this general optimzation may therefore be "incorrect" or at least needs more context about how the flags are being used, because it doesn't hold for any of the basic instructions it called out by itself

Reply:
Actually, it looks like it may partially be that some of the flags for our instructions are incorrect.

Checking both the Intel and AMD manuals
* `or`/`and` should be `Writes_SF`
* `or`/`and`/`xor` should be `WRITES_PF`

Reply:
Was a copy paste error, fixed the table of the 7 existing instructions.

Still think we need to fix to account for the other flags, except `AF` which shouldn't be used for `test` anyways since its undefined.

Reply:
It seemed as if the code wanted to check the minimum possible requirements for skipping the ```test``` instructions. Hence it had a check for just ```OF``` and ```CF```. But this condition returned true for ```bextr``` as well which was not intended. Hence, I just added one more check to exclude ```bextr``` from this list. So now, the instructions which would skip generating ```test``` would be ```or```, ```xor``` and ```and```. In future, if we add instructions which fall in this scenario, we need to further refine this check to exclude the new instructions. But as of now, it seems that the check was intended to keep it at the minimal possible filter. I can still add all other conditions but the filter would still filter out the same instructions. Shall I still add the checks for ZF/PF?

Reply:
True. Looks like we would see this issue with ```andn```, ```popcnt``` as well. will refine the filter further to make sure we are only skipping ```test``` for ```and```, ```or``` and, ```xor```.